### PR TITLE
Enforce object literal shorthand syntax with ESLint

### PR DESCRIPTION
**Pull request by astorije**
All changes in this commit were made with `npm run lint:js -- --fix`.

### CODE REVIEW

**Comment by creesch** [comment]
Wouldn't it be better to make this and issue and do a pull request for this after most of the current pull requests have been merged for 3.0 ? Seems like it will only cause a bunch of merge conflicts either on this end or current pull requests that are otherwise ready to go. 

Also, personally I find this less readable as far as code goes as you basically have things not as neatly aligned anymore in object declarations but more of a zig zag.

**Comment by McInkay** [comment]
I don't think this does make it any less readable, as I don't see how they are less aligned? They are just half the width now.

But yeah, it's going to cause lots of merge conflicts. The problem is, there is never a point where it'll not cause them :-\

**Comment by creesch** [comment]
Well without the shorthand all your objects look like this
```javascript 

const object = {
    var1: value1,
    var2: value2,
    var3: var3,
    var4: value4,
}

```

Which now turns into 

```javascript 

const object = {
    var1: value1,
    var2: value2,
    var3,
    var4: value4,
}
```

So it is no longer a given you can see the value on the right. It is a minor thing. 


As far as preventing merge conflicts, I get that. 

Dunno, if I had known this was coming I'd already followed this style in the PR for settings sync. So you could basically do something like this

- Do `eslint --fix` in 3.0 as the last PR. 
- In the review process for any PR that will be for a version after 3.0 ask people to make any code change with that in mind. 

Personally, having to do a lot of merge conflict solving on a big enough PR always leaves me a bit uneasy about the state of code certainly when the PR has been sitting for a while making me go back and forth a bunch of times making sure I actually merge the correct bits of code. 

You are right though, you have to pick some point in time if you want to make changes like this so might as well do it now. You could pick a minor release but then you probably would still have the same issues with PRs ready for the bigger release after the minor one.

**Comment by McInkay** [comment]
> So it is no longer a given you can see the value on the right. It is a minor thing.

Oh yeah, I see what you mean.

I think this is one of thsoe things that maybe we don't want to enforce everywhere, as there are some examples like this that would not benefit from it, but the majority of cases, we just name them all what the variables were called, and so 

```
var test = {
	var1,
	var2
}
```

definitely looks better than

```
var test = {
	var1: var1,
	var2: var2
}
```

So maybe this is the same sort of thing, unnecessary to enforce, but we can just be improving slowly as we go along? It would also solve the merge conflicts issue.

**Comment by creesch** [comment]
Doesn't eslint allow for more fine-grained control here (it sometimes does), as a start it could be moved to warn.

**Comment by creesch** [comment]
Apparently it does, there is '"consistent-as-needed"' as option.

https://eslint.org/docs/rules/object-shorthand

**Comment by astorije** [comment]
If we wait until next release to merge this kind of PRs, we'll have the same issue. After all, we just did a release and it's already a problem :)

Anyway, I'm totally fine waiting to merge this PR until we merge a bunch of others as it's dead easy to fix conflicts here (it's auto-fixable), but I'm not going to stage ESLint settings, because that would still cause conflicts, only several times (also I don't think it's time well spent).

Note if anyone runs into conflicts after this PR gets merged and want a super easy recipe to fix: ignore all the upstream changes (i.e. keep your version of the code), and run `npm run lint:js -- --fix`. Free conflict resolution :)

**Comment by McInkay** [comment]
What do you think about using consistent-as-needed for this, @astorije ? Would be better, imo.

**Comment by astorije** [comment]
I don't see how it's better actually, I'd rather keep `always`.
Sometimes you are introducing one key/value that requires expansion, and `consistent-as-needed` would force rewriting all the keys, so not a fan.

**Comment by McInkay** [comment]
Well, the point is that these eslint rules we use to make the code look better and more readable. This is why this rule is being enforced, correct? Removing unnecessary characters and making it easier to read some object creation.

The argument for consistent-as-needed is that if an object has 1 or more that needs to be expanded, it is more readable for all of them to be expanded, rather than having to run your eyes in multiple directions more than needed. In this case, the expanded characters are no unnecessary characters, but are helping the readability.

**Comment by astorije** [comment]
> The argument for consistent-as-needed is that if an object has 1 or more that needs to be expanded, it is more readable for all of them to be expanded, rather than having to run your eyes in multiple directions more than needed. In this case, the expanded characters are no unnecessary characters, but are helping the readability.

This is what I disagree with. I find `myFunction({ foo, bar: bad, baz })` more readable than `myFunction({ foo: foo, bar: bad, baz: baz })`.

**Comment by xPaw** [comment]
@astorije I find it more confusing to mix it. I say leave the rule to allow it, and we can clean up code to remove the mixing.

**Comment by creesch** [comment]
To be honest I wouldn't write it like that. 

```javascript
myFunction({ 
    foo, 
    bar: bad, 
    baz 
})
```
And 

```javascript
myFunction({ 
    foo: foo, 
    bar: bad, 
    baz: baz 
})
```

Are usually how these functions are written out for readability and there the second one is much clearer. 
Also in the first example it takes a bit more effort to parse and when not paying attention `bad` could be mistaken for fifth parameter rather than the value for `bar`.

**Comment by astorije** [comment]
Obviously it was on a single line for the sake of commenting here...

> there the second one is much clearer.

Again, I disagree with this. It is very subjective, it's not like one of us holds the truth :)

> Also in the first example it takes a bit more effort to parse and when not paying attention `bad` could be mistaken for fifth parameter rather than the value for `bar`.

That is a bit reaching, IMO. The advantage of this notation is calling out differences, so that `bad` and `bar` cannot be mistaken.

Putting it in practical matters, I find [this](https://github.com/thelounge/lounge/blob/3e11d9a9c1548faa6020e3ea12ca52ee5ae03c37/client/js/constants.js#L108-L115) much more readable/parseable than [this](https://github.com/thelounge/lounge/blob/3e11d9a9c1548faa6020e3ea12ca52ee5ae03c37~/client/js/constants.js#L108-L115), cognitive overflow.

**But here is the kicker:** `consistent-as-needed` is not auto-fixable, so it's an all-or-nothing type of deal :)

**Comment by creesch** [comment]
> Obviously it was on a single line for the sake of commenting here...

Not obvious to me. 

> Again, I disagree with this. It is very subjective, it's not like one of us holds the truth :)

Well yeah, a lot of this is but so far we have provided reasoning and an explanation for how we see it not just "I disagree". 

> That is a bit reaching, IMO. 

In a single line instance? I don't think so, one of the most well known jokes about debuging code is that silly comma or other single character you keep overlooking. When there is a clear pattern of `attribute: value` it is easier to look at glance at something than when that pattern is absent and it sometimes might just be the attribute and other times it might the attribute and value. 

 > consistent-as-needed is not auto-fixable, so it's an all-or-nothing type of deal 

Fair enough, guess that makes it an easy call.

**Comment by astorije** [comment]
> so far we have provided reasoning and an explanation for how we see it not just "I disagree".

Yes, and I did that too 😉

**Comment by astorije** [comment]
> I'd rather we clean up the code so that there is no mixing, but that can be done in the future.

I essentially did that on a bunch of places. Thoughts?

**Comment by McInkay** [comment]
> I essentially did that on a bunch of places. Thoughts?

What do you mean by that? This PR just forces everything to be shorthand if it can be, but it doesn't turn anything from different names into the same name so it can be shorthanded.

For example, before this PR we had stuff like:

```
return {
  chan: chan.id,
  messages: messages,
};
```

Your PR turns it into 

```
return {
  chan: chan.id,
  messages,
};
```

What @xPaw Is saying is that it would be better to have:

```
return {
  chan,
  messages,
};
```

overall (bad example, I know, because we already have a chan in this code, but ... let it be for now).

Now, obviously, we can't just change everything in 1 go, because there are a lot of things (like the example I gave above) which can't just be renamed, due to existing variables. But we don't like the mixed nature of the middle example, and so would prefer if, for this PR, we changed it to `consistent-as-needed` so that we get the 3rd option when we can, but when we can't just now, we get the top, as many of us find it more readable. Eventually, we will change everything to be the 3rd, but only as we edit the code around it.

**Comment by astorije** [comment]
> it doesn't turn anything from different names into the same name so it can be shorthanded.

Yes, I've actually updated this PR to do exactly this, [for example here](https://github.com/thelounge/thelounge/pull/2073/files#diff-386df29cd51cc4ce978bd6fee66927ca), but not everywhere, granted, indeed because:

> we can't just change everything in 1 go, because there are a lot of things (like the example I gave above) which can't just be renamed, due to existing variables.

But where it was very sensible to do so, I did :)

> for this PR, we changed it to `consistent-as-needed`

As mentioned above, my main concern is that `consistent-as-needed` (and `consistent`) is not auto-fixable with the `--fix` flag, which means that this will always have to be fixed by hand, including when other PRs will have to rebase on this.
If it was auto-fixable, I would have made the switch already, but I think that, when in doubt, we should go for the auto-fixable solution as it benefits the many IMO. What do you think about this?

**Comment by astorije** [comment]
@McInkay, did you get any chance to look at my last comment? :)

**Comment by McInkay** [comment]
Yeah, my answer to that would be "meh".

I get the issue, and see why you wouldn't want to do that.

But that would lead me to say I'm not a huge fan of this PR at all in that case. You know that I'm not a fan with enforcing too much with eslint. I like the stuff that can cause real programming errors (`==` vs `===`) and to keep certain things consistent (indenting mainly). But some stuff like this ... just feels unnecessary to force, especially when it's less than ideal.

But whatever, I don't really care enough. I've approved it already, just go for it if you both want it.

**Comment by McInkay** [comment]
Fine with methods.

**Comment by astorije** [comment]
Alright, I think we're getting to a compromise here! *But* I did quite a bit of cleanup for this and I would not feel too great to blindly revert all that.
So what about:

- We enforce `methods`
- We don't enforce `properties`
- I keep the cleanup on things such as:
  ```diff
   function settingSetEmit(name, value) {
  -	socket.emit("setting:set", {
  -		name: name,
  -		value: value,
  -	});
  +	socket.emit("setting:set", {name, value});
   }
  
  -const template = $(templates.msg_preview({preview: preview}));
  +const template = $(templates.msg_preview({preview}));
  ```
  These are no-brainers to me (but again, no enforcing) :)
  - I revert the questionable "cleanups", i.e. those where there is an inconsistency between expanded/shorthand, those requiring subpar gaming-the-system kind of thing (there is a `chan = chan.id` somewhere to please the linter...), etc. so that what you guys find less than ideal doesn't make it.

Deal? :)

**Comment by McInkay** [comment]
Yeah sure, that seems like a nice balance.

**Comment by astorije** [comment]
@McInkay, good to go I think. Makes for a shorter PR, less likely to conflict.

File: client/js/sorting.js [hunk 1]
```diff
@@ -17,18 +17,15 @@ module.exports = function() {
 		forcePlaceholderSize: true,
 		tolerance: "pointer", // Use the pointer to figure out where the network is in the list
 
-		update: function() {
+		update() {
+			const type = "networks";
```
Comment (xPaw) [inline client/js/sorting.js]: This PR looks good now, except for this (and line 44). It feels somewhat unnatural to do it like this.

Comment (astorije) [reply client/js/sorting.js]: Ok, reverted both.

**Pull request merged**